File: manhole/src/main/java/com/freesith/manhole/ui/MockListLayout.java
package com.freesith.manhole.ui;

import android.content.Context;
import android.graphics.Color;
import android.os.Build;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.freesith.manhole.Mox;
import com.example.mox.R;
import com.freesith.manhole.db.bean.Case;
import com.freesith.manhole.db.bean.Flow;
import com.freesith.manhole.db.bean.Mock;
import com.freesith.manhole.db.bean.MockChoice;
import com.freesith.manhole.ui.adapter.CaseAdapter;
import com.freesith.manhole.ui.adapter.ChoiceAdapter;
import com.freesith.manhole.ui.adapter.FlowAdapter;
import com.freesith.manhole.ui.adapter.MockAdapter;
import com.freesith.manhole.ui.interfaces.MonitorListener;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;

public class MockListLayout extends LinearLayout implements View.OnClickListener, FlowAdapter.FlowListener, CaseAdapter.CaseListener, MockAdapter.MockListener, ChoiceAdapter.ChoiceListener {

    private Context context;

    private RecyclerView rvMock;
    private MockAdapter mockAdapter;
    private CaseAdapter caseAdapter;
    private FlowAdapter flowAdapter;
    private ChoiceAdapter enableAdapter;

    private TextView tvEnable;
    private TextView tvFlow;
    private TextView tvCase;
    private TextView tvMock;

    private MonitorListener monitorListener;


    public MockListLayout(Context context) {
        super(context);
        init(context);
    }

    public MockListLayout(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public MockListLayout(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    public MockListLayout(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init(context);
    }


    public void setMonitorListener(MonitorListener monitorListener) {
        this.monitorListener = monitorListener;
    }

    private void init(Context context) {
        this.context = context;
        View view = LayoutInflater.from(context).inflate(R.layout.layout_mock_list, this);

        rvMock = view.findViewById(R.id.rvMock);
        tvFlow = view.findViewById(R.id.tvFlow);
        tvCase = view.findViewById(R.id.tvCase);
        tvMock = view.findViewById(R.id.tvMock);
        tvEnable = view.findViewById(R.id.tvEnable);

        tvFlow.setOnClickListener(this);
        tvCase.setOnClickListener(this);
        tvMock.setOnClickListener(this);
        tvEnable.setOnClickListener(this);

        rvMock.setLayoutManager(new LinearLayoutManager(context));
    }

    public void show() {
        showEnable();
        updateButtons(tvEnable);
    }

    private void showEnable() {
        if (enableAdapter == null) {
            enableAdapter = new ChoiceAdapter(context);
            enableAdapter.setChoiceListener(this);
        }
        ConcurrentHashMap<String, LinkedList<MockChoice>> enableMockMap = Mox.getInstance().enableMockMap;
        List<MockChoice> choiceList = new ArrayList<>();
        if (!enableMockMap.isEmpty()) {
            for(LinkedList linkedList : enableMockMap.values()) {
                choiceList.addAll(linkedList);
            }
        }
        enableAdapter.setList(choiceList);
        rvMock.setAdapter(enableAdapter);
        enableAdapter.notifyDataSetChanged();

    }

    private void showFlow() {
        if (flowAdapter == null) {
            flowAdapter = new FlowAdapter(context);
            flowAdapter.setFlowListener(this);
        }
        List<Flow> flows = Mox.getInstance().getFlows();
        flowAdapter.setList(flows);
        rvMock.setAdapter(flowAdapter);
        flowAdapter.notifyDataSetChanged();
    }

    private void showCase() {
        if (caseAdapter == null) {
            caseAdapter = new CaseAdapter(context);
            caseAdapter.setCaseListener(this);
        }
        List<Case> cases = Mox.getInstance().getCases();
        caseAdapter.setList(cases);
        rvMock.setAdapter(caseAdapter);
        caseAdapter.notifyDataSetChanged();
    }

    private void showMock() {
        if (mockAdapter == null) {
            mockAdapter = new MockAdapter(context);
            mockAdapter.setMockListener(this);
        }
        List<Mock> mocks = Mox.getInstance().getMocks();
        mockAdapter.setList(mocks);
        rvMock.setAdapter(mockAdapter);
        mockAdapter.notifyDataSetChanged();
    }


    @Override
    public void onClick(View v) {
        if (v.getId() == R.id.tvFlow) {
            showFlow();
            updateButtons(v);
        } else if (v.getId() == R.id.tvCase) {
            showCase();
            updateButtons(v);
        } else if (v.getId() == R.id.tvMock) {
            showMock();
            updateButtons(v);
        } else if (v.getId() == R.id.tvEnable) {
            showEnable();
            updateButtons(v);
        }


    }

    private void updateButtons(View v) {
        tvCase.setBackgroundColor(v.getId() == R.id.tvCase ? Color.WHITE : Color.TRANSPARENT);
        tvEnable.setBackgroundColor(v.getId() == R.id.tvEnable ? Color.WHITE : Color.TRANSPARENT);
        tvFlow.setBackgroundColor(v.getId() == R.id.tvFlow ? Color.WHITE : Color.TRANSPARENT);
        tvMock.setBackgroundColor(v.getId() == R.id.tvMock ? Color.WHITE : Color.TRANSPARENT);
    }

    @Override
    public void onFlowEnableChanged(String name, boolean enable, int position) {
        Flow flow = flowAdapter.get(position);
        if (flow != null) {
            Mox.getInstance().updateFlowEnable(name, enable);
            flow.enable = enable;
        }
    }

    @Override
    public void onCaseEnableChanged(String name, boolean enable, int position) {
        Case caze = caseAdapter.get(position);
        if (caze != null) {
            Mox.getInstance().updateCaseEnable(name, enable);
            caze.enable = enable;
        }
    }

    @Override
    public void onMockEnableChanged(String name, boolean enable, int position) {
        Mock mock = mockAdapter.get(position);
        if (mock != null) {
            Mox.getInstance().updateMockEnable(name, enable);
            mock.enable = enable;
        }
    }

    @Override
    public void onMockClick(Mock mock) {
        if (monitorListener != null) {
            monitorListener.onShowSingleMock(mock);
        }
    }

    @Override
    public void onChoiceEnableChanged(MockChoice choice, boolean enable, int position) {
        Mox.getInstance().updateMockChoiceEnable(choice.mockName, choice.index, enable);
    }

    @Override
    public void onChoiceClick(MockChoice mock) {

    }
}

File: manhole/src/main/java/com/freesith/manhole/ui/MockListLayout.kt
package com.freesith.manhole.ui

import android.content.Context
import android.graphics.Color
import android.os.Build
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.FrameLayout
import android.widget.LinearLayout
import androidx.annotation.RequiresApi
import androidx.recyclerview.widget.LinearLayoutManager
import com.freesith.manhole.ManholeMock
import com.freesith.manhole.R
import com.freesith.manhole.bean.Mock
import com.freesith.manhole.bean.MockChoice
import com.freesith.manhole.ui.adapter.CaseAdapter
import com.freesith.manhole.ui.adapter.CaseAdapter.CaseListener
import com.freesith.manhole.ui.adapter.EnableChoiceAdapter
import com.freesith.manhole.ui.adapter.EnableChoiceAdapter.ChoiceListener
import com.freesith.manhole.ui.adapter.FlowAdapter
import com.freesith.manhole.ui.adapter.FlowAdapter.FlowListener
import com.freesith.manhole.ui.adapter.MockAdapter
import com.freesith.manhole.ui.adapter.MockAdapter.MockListener
import com.freesith.manhole.ui.interfaces.MonitorListener
import com.freesith.manhole.ui.util.ViewUtil
import kotlinx.android.synthetic.main.layout_mock_list.view.*
import java.util.*

class MockListLayout : LinearLayout, View.OnClickListener, FlowListener, CaseListener,
    MockListener, ChoiceListener {
    private var thisContext: Context? = null
    private var mockAdapter: MockAdapter? = null
    private var caseAdapter: CaseAdapter? = null
    private var flowAdapter: FlowAdapter? = null
    private var enableAdapter: EnableChoiceAdapter? = null
    private var monitorListener: MonitorListener? = null

    constructor(context: Context) : super(context) {
        init(context)
    }

    constructor(
        context: Context,
        attrs: AttributeSet?
    ) : super(context, attrs) {
        init(context)
    }

    constructor(
        context: Context,
        attrs: AttributeSet?,
        defStyleAttr: Int
    ) : super(context, attrs, defStyleAttr) {
        init(context)
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    constructor(
        context: Context,
        attrs: AttributeSet?,
        defStyleAttr: Int,
        defStyleRes: Int
    ) : super(context, attrs, defStyleAttr, defStyleRes) {
        init(context)
    }

    fun setMonitorListener(monitorListener: MonitorListener?) {
        this.monitorListener = monitorListener
    }

    private fun init(context: Context) {
        this.thisContext = context
        val view =
            LayoutInflater.from(context).inflate(R.layout.layout_mock_list, this)
        tvFlow.setOnClickListener(this)
        tvCase.setOnClickListener(this)
        tvMock.setOnClickListener(this)
        tvEnable.setOnClickListener(this)
        rvMock.setLayoutManager(LinearLayoutManager(context))
    }

    fun show() {
        showEnable()
        updateButtons(tvEnable)
    }

    private fun showEnable() {
        if (enableAdapter == null) {
            enableAdapter = EnableChoiceAdapter(thisContext)
            enableAdapter!!.setChoiceListener(this)
        }
        val enableMockMap = ManholeMock.enableMockMap
        val choiceList: MutableList<MockChoice> =
            ArrayList()
        if (!enableMockMap.isEmpty()) {
            for (linkedList in enableMockMap.values) {
                choiceList.addAll(linkedList)
            }
        }
        enableAdapter!!.setList(choiceList)
        rvMock!!.adapter = enableAdapter
        enableAdapter!!.notifyDataSetChanged()
    }

    private fun showFlow() {
        if (flowAdapter == null) {
            flowAdapter = FlowAdapter(thisContext)
            flowAdapter!!.setFlowListener(this)
        }
        val flows = ManholeMock.flows
        flowAdapter!!.setList(flows)
        rvMock!!.adapter = flowAdapter
        flowAdapter!!.notifyDataSetChanged()
    }

    private fun showCase() {
        if (caseAdapter == null) {
            caseAdapter = CaseAdapter(thisContext)
            caseAdapter!!.setCaseListener(this)
        }
        val cases = ManholeMock.cases
        caseAdapter!!.setList(cases)
        rvMock!!.adapter = caseAdapter
        caseAdapter!!.notifyDataSetChanged()
    }

    private fun showMock() {
        if (mockAdapter == null) {
            mockAdapter = MockAdapter(thisContext)
            mockAdapter!!.setMockListener(this)
        }
        val mocks = ManholeMock.mocks
        mockAdapter!!.setList(mocks)
        rvMock!!.adapter = mockAdapter
        mockAdapter!!.notifyDataSetChanged()
    }

    override fun onClick(v: View) {
        if (v.id == R.id.tvFlow) {
            showFlow()
            updateButtons(v)
        } else if (v.id == R.id.tvCase) {
            showCase()
            updateButtons(v)
        } else if (v.id == R.id.tvMock) {
            showMock()
            updateButtons(v)
        } else if (v.id == R.id.tvEnable) {
            showEnable()
            updateButtons(v)
        }
    }

    private fun updateButtons(v: View?) {
        tvCase!!.setBackgroundColor(if (v!!.id == R.id.tvCase) Color.WHITE else Color.TRANSPARENT)
        tvEnable!!.setBackgroundColor(if (v.id == R.id.tvEnable) Color.WHITE else Color.TRANSPARENT)
        tvFlow!!.setBackgroundColor(if (v.id == R.id.tvFlow) Color.WHITE else Color.TRANSPARENT)
        tvMock!!.setBackgroundColor(if (v.id == R.id.tvMock) Color.WHITE else Color.TRANSPARENT)
    }

    override fun onFlowEnableChanged(
        name: String,
        enable: Boolean,
        position: Int
    ) {
        val flow = flowAdapter!![position]
        if (flow != null) {
            ManholeMock.updateFlowEnable(name, enable)
            flow.enable = enable
        }
    }

    override fun onFlowClick(name: String) {
        val caseLayout = CaseLayout(thisContext)
        ViewUtil.findCoverLayout(this).addView(
            caseLayout,
            FrameLayout.LayoutParams.MATCH_PARENT,
            FrameLayout.LayoutParams.MATCH_PARENT
        )
        caseLayout.showFlow(name)
    }

    override fun onCaseEnableChanged(
        name: String,
        enable: Boolean,
        position: Int
    ) {
        val caze = caseAdapter!![position]
        if (caze != null) {
            ManholeMock.updateCaseEnable(name, enable)
            caze.enable = enable
        }
    }

    override fun onCaseClick(name: String) {
        val caseLayout = CaseLayout(thisContext)
        ViewUtil.findCoverLayout(this).addView(
            caseLayout,
            FrameLayout.LayoutParams.MATCH_PARENT,
            FrameLayout.LayoutParams.MATCH_PARENT
        )
        caseLayout.showCase(name)
    }

    override fun onMockEnableChanged(
        name: String,
        enable: Boolean,
        position: Int
    ) {
        val mock = mockAdapter!![position]
        if (mock != null) {
            ManholeMock.updateMockEnable(name, enable)
            mock.enable = enable
        }
    }

    override fun onMockClick(mock: Mock) {
        if (monitorListener != null) {
            monitorListener!!.onShowSingleMock(mock)
        }
    }

    override fun onChoiceEnableChanged(
        choice: MockChoice,
        enable: Boolean,
        position: Int
    ) {
        ManholeMock.updateMockChoiceEnable(choice.mockName, choice.index, enable)
    }

    override fun onMockNameClick(name: String) {
        val mockView = MockView(thisContext)
        val mock = ManholeMock.findMockByName(name)
        if (mock != null) {
            ViewUtil.findCoverLayout(this).addView(
                mockView,
                FrameLayout.LayoutParams.MATCH_PARENT,
                FrameLayout.LayoutParams.MATCH_PARENT
            )
            mockView.showMock(mock)
        }
    }

    override fun onChoiceClick(mock: MockChoice) {}
}